Marlin/src/HAL/STM32/Sd2Card_sdio_stm32duino.cpp

```diff
@@ -88,18 +88,37 @@
 
     MKS Robin board seems to have stable SDIO with BusWide 1bit and ClockDiv 8 i.e. 4.8MHz SDIO clock frequency
     Additional testing is required as there are clearly some 4bit initialization problems
-
-    Add -DTRANSFER_CLOCK_DIV=8 to build parameters to improve SDIO stability
   */
 
-  #ifndef TRANSFER_CLOCK_DIV
-    #define TRANSFER_CLOCK_DIV (uint8_t(SDIO_INIT_CLK_DIV) / 40)
-  #endif
-
   #ifndef USBD_OK
     #define USBD_OK 0
   #endif
 
+  // Target Clock, configurable. Default is 18MHz, from STM32F1
+  #ifndef SDIO_CLOCK
+    #define SDIO_CLOCK                         18000000       /* 18 MHz */
+  #endif
+
+  // SDIO retries, configurable. Default is 3, from STM32F1
+  #ifndef SDIO_READ_RETRIES
+    #define SDIO_READ_RETRIES                  3
+  #endif
+
+  // SDIO Max Clock (naming from STM Manual, don't change)
+  #define SDIOCLK 48000000
+
+  static uint32_t sdio_set_clock(uint32_t clk) {
+    /*
+     * limit the SDIO master clock to 8/3 of PCLK2. See STM32 Manuals
+     * Also limited to no more than 48Mhz (SDIOCLK).
+     */
+    const uint32_t pclk2 = HAL_RCC_GetPCLK2Freq();
+    clk = min(clk, ((uint32_t)pclk2/3)*8);
```

Comment:
You should get more accurate results if you multiply before divide.

For example: If `pclk2 = 100`, ideal answer is 266.66667...

`(uint32_t)pclk2 / 3 * 8 == 264`

`(uint32_t)pclk2 * 8 / 3 == 266`

Ideal answer is 266.66

Reply:
I'm following the STM32 manual... I know that the order don't make too much diference, but I'm following the order of stm32 manual...

Reply:
Where in the manual? I would like to look at it.

I would consider this a math bug, right now.

Reply:
My point is that the order _does_ make a difference. When doing integer math operations need to be ordered to minimize rounding errors.

Reply:
<img width="1057" alt="Screen Shot 2020-12-26 at 18 01 20" src="https://user-images.githubusercontent.com/81722/103159139-76e72300-47a4-11eb-979b-93410ffab362.png">

Reply:
If / 3 * 8 gives a lower value, I think we should keep it.

Reply:
Which manual does this come from? I would like to look at it in context.

The image you show is not code. It does not necessary assume integer rounding will occur.

Reply:
F4, page 1022: https://www.st.com/resource/en/reference_manual/dm00031020-stm32f405-415-stm32f407-417-stm32f427-437-and-stm32f429-439-advanced-arm-based-32-bit-mcus-stmicroelectronics.pdf

F1, page 569: https://www.st.com/resource/en/reference_manual/cd00171190-stm32f101xx-stm32f102xx-stm32f103xx-stm32f105xx-and-stm32f107xx-advanced-arm-based-32-bit-mcus-stmicroelectronics.pdf

Reply:
I found the manual. Reviewing now.

Reply:
So, if you treated their formula as integer math:

`Frequency(PCLK2) >= 0 * Frequency(SDIO_CK)`

This would always evaluate to true!
`Frequency(PCLK2) >= 0`

That is clearly nonsense, meaning they are not providing an integer math formula. I am sure their intention is more like:
`Frequency(PCLK2) >= 0.375 * Frequency(SDIO_CK)`

Reply:
what is the verdict?

Reply:
For me, SDIO_CK must be less than 8/3 of PCLK2.

Reply:
Or, SDIO_CK must be less than 0.375 of PCLK2, if we don't like fractions :-)

Reply:
Well, that is backwards. 

SDIO_CK must be less than 8/3 of PCLK2.

The integer math copied from Maple is wrong. The fix is simple.

` clk = min(clk, ((uint32_t)pclk2/3)*8);`

to

` clk = min(clk, ((uint32_t)pclk2 * 8 / 3);`

Reply:
`clk = min(clk, (uint32_t)(pclk2 * 8 / 3));`?

Reply:
`pclk2` is already a `uint32_t`, so the cast and parenthesis aren't needed at all. I didn't notice that before.

`clk = min(clk, pclk2 * 8 / 3);